File: python-codes/03_DFS_BFS/bfs.py
# BFS - Breadth First Search
from collections import deque

def bfs(graph, i, visited):
  visited[i] = True
  queue = deque([i])

# queue 가 빌때까지 반복
  while queue:
    v = queue.popleft()
    print(v, end=' ')

    for j in graph[v]:
      if not visited[j]:
        queue.append(j)
        visited[j] = True

graph = [
  [],
  [2, 3, 8],
  [1, 7],
  [1, 4, 5],
  [3, 5],
  [3, 4],
  [7],
  [2, 6, 8],
  [1, 7]
]

visited = [False] * 9

print(bfs(graph, 1, visited))
# 1, 2, 3, 8, 7, 4, 5, 6
# 1 2 3 8 7 4 5 6 None

File: java-codes/src/practice/algorithm/bfsdfs/Bfs.java
package practice.algorithm.bfsdfs;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

/**
 * bfs
 * 1 2 3 8 7 4 5 6
 * */
public class Bfs {
    public static void main(String[] args) {
        int[][] graph = new int[][]{
                {},
                {2, 3, 8},
                {1, 7},
                {1, 4, 5},
                {3, 5},
                {3, 4},
                {7},
                {2, 6, 8},
                {1, 7}
        };

        boolean[] visited = new boolean[9];
        Arrays.fill(visited, false);

        bfs(graph, visited, 1);
    }

    private static void bfs(int[][] graph, boolean[] visited, int i) {
        visited[i] = true;

        Queue<Integer> queue = new LinkedList<>();
        queue.add(i);

        while (!queue.isEmpty()) {
            Integer v = queue.poll();
            System.out.print(v + " ");

            for (int node : graph[v]) {
                if (!visited[node]) {
                    visited[node] = true;
                    queue.add(node);
                }
            }
        }
    }

}
